source/s3/s3_test.go

```diff
@@ -0,0 +1,184 @@
+package s3
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"testing"
+
+	"github.com/MiLk/nsscache-go/cache"
+	"github.com/stretchr/testify/assert"
+)
+
+func TestS3Source_FillPasswdCache1(t *testing.T) {
+	dir, err := ioutil.TempDir("/tmp", "nsscache-go-")
+	assert.Nil(t, err)
+	defer os.RemoveAll(dir)
+
+	r := `[{
+	"name": "foo",
+	"passwd": "x",
+	"uid": 1000,
+	"gid": 1000,
+	"gecos": "Mr Foo",
+	"dir": "/home/foo",
+	"shell": "/bin/bash"
+},
+{
+	"name": "var",
+	"passwd": "x",
+	"uid": 1001,
+	"gid": 1000,
+	"gecos": "Mr Var",
+	"dir": "/home/var",
+	"shell": "/bin/bash"
+}]`
+	svc := CreateMockS3GetObjectClient(r, nil)
+	prefix := fmt.Sprintf("secret/%s", "nsscache-test")
+	src := CreateS3Source(svc, prefix, "testing-bucket")
+	c := cache.NewCache()
+
+	assert.Nil(t, src.FillPasswdCache(c))
+
+	var b bytes.Buffer
+	n, err := c.WriteTo(&b)
+
+	assert.Nil(t, err)
+	assert.EqualValues(t, 86, n)
+	expected := `foo:x:1000:1000:Mr Foo:/home/foo:/bin/bash
+var:x:1001:1000:Mr Var:/home/var:/bin/bash
+`
+	assert.Equal(t, expected, b.String())
+}
+
+func TestS3Source_FillPasswdCache2(t *testing.T) {
+	// json decoding error
+	dir, err := ioutil.TempDir("/tmp", "nsscache-go-")
+	assert.Nil(t, err)
+	defer os.RemoveAll(dir)
+
+	r := `[{
+	"name": "foo",
+	"passwd": "x,
+	"uid": 1000,
+	"gid": 1000,
+	"gecos": "Mr Foo",
+	"dir": "/home/foo",
+	"shell": "/bin/bash"
+}]`
+	svc := CreateMockS3GetObjectClient(r, nil)
+	prefix := fmt.Sprintf("secret/%s", "nsscache-test")
+	src := CreateS3Source(svc, prefix, "testing-bucket")
+	c := cache.NewCache()
+
+	err = src.FillPasswdCache(c)
+	expectedErr := "json decoding: invalid character '\\n' in string literal"
+	assert.Equal(t, expectedErr, err.Error())
+}
+
+func TestS3Source_FillPasswdCache3(t *testing.T) {
+	// json does not match entry format
+	dir, err := ioutil.TempDir("/tmp", "nsscache-go-")
+	assert.Nil(t, err)
+	defer os.RemoveAll(dir)
+
+	r := `[{
+	"name": "Foo",
+	"passwd": "x",
+	"uid": "1000",
+	"gid": 1000,
+	"gecos": "Mr Foo",
+	"dir": "/home/foo",
+	"shell": "/bin/bash"
+}]`
+	svc := CreateMockS3GetObjectClient(r, nil)
+	prefix := fmt.Sprintf("secret/%s", "nsscache-test")
+	src := CreateS3Source(svc, prefix, "testing-bucket")
+	c := cache.NewCache()
+
+	err = src.FillPasswdCache(c)
+	expectedErr := "json does not match entry format: json: cannot unmarshal string into Go struct field PasswdEntry.uid of type uint32"
+	assert.Equal(t, expectedErr, err.Error())
+}
+
+func TestS3Source_FillPasswdCache4(t *testing.T) {
```

Comment:
You can specify the name you want in the function name.
It can be `TestS3Source_FillPasswdCache_DownloadError`